### PR TITLE
Make new submission notifications configurable

### PR DESCRIPTION
**Pull request by northeastprince**

### CODE REVIEW

**Review by flexcommits — APPROVED** [review]
LGTM!!!

File: app/models/user/settings.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+User::DEFAULT_SETTINGS = {
+  new_submission_notifications: true
+}
+
+module User::Settings
+  extend ActiveSupport::Concern
+
+  included do
+    scope :with_setting_not_disabled, ->(setting) { where("settings ->> ? IS NULL OR settings ->> ? != 'false'", setting.to_s, setting.to_s) }
```
Comment (garyhtou) [inline app/models/user/settings.rb:13]: The current code assumes `NULL` to be a falsy value. In reality, we likely want `NULL` to be the default value.

---

I think we might want:

```suggestion
    scope :with_setting_enabled, ->(setting) { where("COALESCE(settings ->> ?, ?) == 'true'", setting.to_s, User::DEFAULT_SETTINGS[setting]) }
```

or

```suggestion
    scope :with_setting, ->(setting, value) { where("COALESCE(settings ->> ?, ?) == ?", setting.to_s, User::DEFAULT_SETTINGS[setting], value) }
```

I haven't tested this code, but `COALESCE` should allow us to apply the default settings even if the database value is `NULL`.

---

side note: I wonder if `User::DEFAULT_SETTINGS[setting]` needs to be serialized as JSONB before being passed into the query above.

File: app/views/hackathons/submission_mailer/admin_notification.html.erb [hunk 2]
```diff
@@ -3,3 +3,7 @@
   <%= link_to @hackathon.name, admin_hackathon_url(@hackathon) %>
   was submitted!
 </p>
+
+<p>
+  Don't wanna get these emails? <%= link_to "Toggle new submission notifications!", admin_submissions_notifications_url %>
+</p>
```
Comment (garyhtou) [inline app/views/hackathons/submission_mailer/admin_notification.html.erb:9]:
```suggestion
<% content_for :unsubscribe do %>
  <p>
    Don't wanna get these emails? <%= link_to "Toggle new submission notifications!", admin_submissions_notifications_url %>
  </p>
<% end %>
```

File: app/views/hackathons/submission_mailer/admin_notification.text.erb [hunk 3]
```diff
@@ -1,3 +1,5 @@
 A new Hackathon named "<%= @hackathon.name %>" was submitted!
 
 View the submission here: <%= admin_hackathon_url(@hackathon) %>
+
+Toggle notifications for new submissions here: <%= admin_submissions_notifications_url %>
```
Comment (garyhtou) [inline app/views/hackathons/submission_mailer/admin_notification.text.erb:5]:
```suggestion
<% content_for :unsubscribe do %>
  Toggle notifications for new submissions here: <%= admin_submissions_notifications_url %>
<% end %>
```

**Review by garyhtou — CHANGES_REQUESTED** [review]
I like that we're doing this! There are a couple functional bugs to handle

**Comment by garyhtou** [comment]
It might also be nice to add some tests around this

**Comment by garyhtou** [comment]
Also, i just came across https://api.rubyonrails.org/classes/ActiveRecord/Store.html

might be handy!

**Comment by northeastprince** [comment]
@garyhtou what would the benefit be vs jsonb?

**Comment by northeastprince** [comment]
And yeah, I'm planning on adding tests once the implementation is sound.

**Comment by garyhtou** [comment]
> @garyhtou what would the benefit be vs jsonb?

I haven't taken a close look into this. it seems like it supports

> dirty tracking methods (key_changed?, key_was and key_change)
https://api.rubyonrails.org/classes/ActiveRecord/Store.html#:~:text=dirty%20tracking%20methods%20(key_changed%3F%2C%20key_was%20and%20key_change)

but doesn't provide scopes. so it might not be the best fit for our needs. just wanted to share it in case you wanted to do a deep dive into it

**Review by garyhtou — APPROVED** [review]
the approach/design looks good to me!

**Pull request merged**